src/Neo/SmartContract/Native/CryptoLib.BN254.cs

```diff
@@ -0,0 +1,203 @@
+            Mcl.mclBnG1_add(ref result, g1x, g1y);
+            Mcl.mclBnG1_normalize(ref result, result);
+
+            return new InteropInterface(new Bn254G1(BN254.SerializeG1(result)));
+        }
+
+        [ContractMethod(Hardfork.HF_Faun, CpuFee = 1 << 5)]
+        public static bool Bn254Equal(InteropInterface x, InteropInterface y)
+        {
+            return (x.GetInterface<object>(), y.GetInterface<object>()) switch
+            {
+                (Bn254G1 a, Bn254G1 b) => a.SequenceEqual(b),
+                (Bn254G2 a, Bn254G2 b) => a.SequenceEqual(b),
+                _ => throw new ArgumentException("BN254 type mismatch")
+            };
+        }
+
+        [ContractMethod(Hardfork.HF_Faun, CpuFee = 1 << 21)]
+        public static InteropInterface Bn254Mul(InteropInterface point, byte[] scalar)
+        {
+            ArgumentNullException.ThrowIfNull(scalar);
+            if (!BN254.TryDeserializeScalar(scalar, out var mul))
+                throw new ArgumentException("Invalid BN254 scalar", nameof(scalar));
+
+            var source = GetBn254G1(point);
+            if (!source.TryGetPoint(out var g1))
+                throw new ArgumentException("Invalid BN254 point data");
+
+            mclBnG1 result = default;
+            Mcl.mclBnG1_mul(ref result, g1, mul);
+            Mcl.mclBnG1_normalize(ref result, result);
+
+            return new InteropInterface(new Bn254G1(BN254.SerializeG1(result)));
+        }
+
+        [ContractMethod(Hardfork.HF_Faun, CpuFee = 1 << 21)]
+        public static byte[] Bn254Pairing(Array pairs)
+        {
+            ArgumentNullException.ThrowIfNull(pairs);
+
+            if (pairs.Count == 0)
+                return BN254.Pairing(System.Array.Empty<byte>());
+
+            byte[] buffer = new byte[pairs.Count * BN254.PairInputLength];
+            for (int i = 0; i < pairs.Count; i++)
+            {
+                if (pairs[i] is not Array pair || pair.Count != 2)
+                    throw new ArgumentException("BN254 pairing pairs must contain g1 and g2 points");
+
+                if (pair[0] is not InteropInterface g1Interface)
+                    throw new ArgumentException("BN254 pairing requires interop points");
+                if (pair[1] is not InteropInterface g2Interface)
+                    throw new ArgumentException("BN254 pairing requires interop points");
+
+                var g1 = GetBn254G1(g1Interface);
+                var g2 = GetBn254G2(g2Interface);
+
+                var g1Bytes = g1.Encoded;
+                var g2Bytes = g2.Encoded;
+                var slice = buffer.AsSpan(i * BN254.PairInputLength, BN254.PairInputLength);
+                g1Bytes.CopyTo(slice[..g1Bytes.Length]);
+                g2Bytes.CopyTo(slice[g1Bytes.Length..]);
+            }
+
+            return BN254.Pairing(buffer);
+        }
+
+        [ContractMethod(Hardfork.HF_Faun, CpuFee = 1 << 19, Name = "bn254_add")]
+        public static byte[] Bn254AddRaw(byte[] input)
```

Comment:
Please, remove `Bn254*Raw` methods. Standard methods that work with InteropInterface are sufficient.